### PR TITLE
PD: Add tests for InvoluteGear

### PR DESCRIPTION
**Pull request by jbaehr**
There are some basic tests to cover the gear profile and its parameters and some integration tests for the interaction with PD Pad and Pocket.

I followed the same pattern as the other PartDesign tests, e.g. not using PEP8 style method names. Eventually some helper methods could be extracted and moved to the common base framework in Mod/Test, but that's another step.

--------
Thank you for creating a pull request to contribute to FreeCAD! To ease integration, we ask you to conform to the following items. Pull requests which don't satisfy all the items below might be rejected. If you are in doubt with any of the items below, don't hesitate to ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10)!

- [x]  Your pull request is confined strictly to a single module. That is, all the files changed by your pull request are either in `App`, `Base`, `Gui` or one of the `Mod` subfolders. If you need to make changes in several locations, make several pull requests and wait for the first one to be merged before submitting the next ones
- [ ]  In case your pull request does more than just fixing small bugs, make sure you discussed your ideas with other developers on the FreeCAD forum
- [ ]  Your branch is [rebased](https://git-scm.com/docs/git-rebase) on latest master `git pull --rebase upstream master`
- [ ]  All FreeCAD unit tests are confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [x]  All commit messages are [well-written](https://chris.beams.io/posts/git-commit/) ex: `Fixes typo in Draft Move command text`
- [x]  Your pull request is well written and has a good description, and its title starts with the module name, ex: `Draft: Fixed typos`
- [ ]  Commit messages include `issue #<id>` or `fixes #<id>` where `<id>` is the [FreeCAD bug tracker issue number](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) in case a particular commit solves or is related to an existing issue on the tracker. Ex: `Draft: fix typos - fixes #0004805`

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.20 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=56135).

---

### CODE REVIEW

**Comment by wwmayer** [comment]
You also have to modify the corresponding CMake file as otherwise the new file won't be copied/installed.

**Comment by jbaehr** [comment]
> You also have to modify the corresponding CMake file as otherwise the new file won't be copied/installed.

Thanks. Didn't notice that as I worked with symlinks and wasn't recompiling when only working on python stuff.

File: src/Mod/PartDesign/PartDesignTests/TestInvoluteGear.py
Comment (luzpaz) [inline src/Mod/PartDesign/PartDesignTests/TestInvoluteGear.py:1]: Line 1 needs a line of `#**************************` identical to Line 20
Line 2: 
`#    Copyright (c) 2022 Jonas Bähr <jonas.baehr@web.de>                          *`

No need to specify LGPL on line 2 because it's mentioned below

**Review by luzpaz — CHANGES_REQUESTED** [review]

File: src/Mod/PartDesign/PartDesignTests/TestInvoluteGear.py
Comment (wwmayer) [reply src/Mod/PartDesign/PartDesignTests/TestInvoluteGear.py:1]: It's a copy&paste issue. As far a I see all PD test scripts use this format of the license text. So, it should be fixed in a separate commit.

**Comment by wwmayer** [comment]
When running the unit tests I get a failure in the function testCustomizedGearProfile. Error message is:
```
TestInvoluteGear.py", line 70, in testCustomizedGearProfile
    self.assertAlmostEqual(outer_diameter, bb.XLength, delta=0.01, msg="Wrong horizontal extend.")
  File "/usr/lib/python3.6/unittest/case.py", line 878, in assertAlmostEqual
    raise self.failureException(msg)
AssertionError: 14 != 13.980091909670271 within 0.01 delta : Wrong horizontal extend.
```

Tests with the (OCC) bounding box expansions are a bit tricky because it's not guaranteed that they exactly fit.

**Comment by jbaehr** [comment]
> Tests with the (OCC) bounding box expansions are a bit tricky because it's not guaranteed that they exactly fit.

I see... I also observed the bounding box slightly smaller than expected but thought this was due to the fact that the gear's top land is generated as a straight line instead of an arc. That's why I used this 0.01 tolerance, which is already way larger than `Precision::Confusion`. Didn't know that OCCs BB is somewhat loose and not portable/reliable.

I'll change this to a test with intersecting circles: Below the root circle and above the tip circle there should not be an intersection, but the pitch circle should. 

 (btw: Are the methods/constants from OCC's Precision header exposed to python somewhere?)

**Comment by wwmayer** [comment]
Here some background information about bounding boxes: https://opencascade.blogspot.com/2009/04/all-inclusive-bounding-box.html

> (btw: Are the methods/constants from OCC's Precision header exposed to python somewhere?)

No, I don't think so. But it's good idea to add them, too. Just wait a little bit...

**Comment by wwmayer** [comment]
... here we are: 20b86e55b8

File: src/Mod/PartDesign/PartDesignTests/TestInvoluteGear.py
Comment (wwmayer) [reply src/Mod/PartDesign/PartDesignTests/TestInvoluteGear.py:1]: With 6461f9f6d2 I made the license text uniform of the test modules. So can you adjust it in your file too, please?

Comment (jbaehr) [reply src/Mod/PartDesign/PartDesignTests/TestInvoluteGear.py]: done

**Comment by jbaehr** [comment]
Thanks for all the replies. I've now changed the bounding box based check to an intersection based one. And thereby already used the new Precision export.

I'm not sure about the internal precision of the `distToShape` method, though. If the results are obtained analytically, then the check with `Precision::Intersection` is fine, maybe even required to ensure tangents with similar curvature are correctly handled. However, if there is some numerical algorithm behind all this, my check for intersection may need refinement. -- Anyway, for the specific application in the gear's test it's irrelevant, but this could become interesting when extracting these helpers and reusing them elsewhere.

**Comment by benj5378** [comment]
I believe that `testInvoluteGear` should be added to __init__.py? I guess it's an issue for all the others modules too...

**Comment by wwmayer** [comment]
@benj5378 
> I believe that testInvoluteGear should be added to __init__.py? I guess it's an issue for all the others modules too...

Why? At least the way it's done the test system can import all needed modules.

**Comment by wwmayer** [comment]
> It's an improvement suggestion. Let me know if I am incorrect.

OK, I see. Yes, it would be an option to adjust __init__.py accordingly. But I will address it after merging the PR.

**Comment by wwmayer** [comment]
f9b37729fe

**Comment by benj5378** [comment]
@wwmayer Cool, that sounds good. I agree that this mistake should not block the PR since it is minor and espacially as its missing in many FreeCAD python modules.

Without being sure, I believe the file should also be named `__init__.py`

**Comment by jbaehr** [comment]
> If you need to use `PartDesignTests`, you currently need to write: `from PartDesignTests import testInvoluteGear, test2, test3, test4, etc`. By creating the `__init__.py` file correctly, all test from `PartDesignTests` can be imported by `import PartDesignTests`
> 

Honestly, I don't see a need to this. Feel free to convince me otherwise, though. Here are my arguments:

First, from my understanding, the python way prefers explicit over implicit. Having a very slim, often empty, `__init__.py` is common practice; even recommended, e.g. by the famous [Hitchhiker’s Guide to Python](https://docs.python-guide.org/writing/structure/#packages) (which gives some rationales for this, too).

Then, for whom does this makes life easier? I mean, who needs to consume test classes other than the test runner. We already to add them to `TestPartDesignApp.py` (or `TestPartDesignGui.py` in the case they need a GUI). I mean, where are all the `import PartDesignTests` statement you need to optimise?

> f9b3772

Now we have again one more place to remember where new tests have to be added, too. I suggest to revert this, unless I've overseen some benefits that justify this double maintenance.

**Comment by wwmayer** [comment]
> Cool, that sounds good. I agree that it should not block the PR since it is minor and espacially as its already missing in many FreeCAD python modules.

I don't think there is anything wrong with leaving the __init__.py files empty. Many other Python packages leave it empty, too.

> Now we have again one more place to remember where new tests have to be added, too. I suggest to revert this, unless I've overseen some benefits that justify this double maintenance.

That's actually a valid point.

After all I don't mind if we have the stuff inside __init__.py or not. But I also agree that I don't see a real benefit to have it.

**Comment by benj5378** [comment]
@jbaehr

I think that you have some good points. Though having done some searches on Google, I don't know if it's considered good practice, it seems more like people have different opinions and it changes from case to case.

I think you have some good points.

**Pull request merged**
